### PR TITLE
26 training regime

### PR DESCRIPTION
**Pull request by philswatton**
This PR has ended up becoming something of a catch-all for not only the training regime, but some general fixes and tweaking of the drop-only experiment group. The changes are:

- After reviewing some papers and following discussion in the stand-ups, it seems it's best to stick to a training regime similar to that in phase 1. The main reason is to allow us to train a large number of models in a relatively short time. The SOTA procedure for ResNet in one of the papers was a 600-epoch procedure, and a shorter 100-epoch procedure in the same paper took 100 hours even with GPU.
- That said, I've decreased the learning rate in line with one of the papers. This did seem to marginally improve things, but this was with two different seeds
- I've added an explicit momentum argument to our resnet model. Before this was set to 0.9 in the code
- Some arguments to the model (weight decay, momentum) are now explicitly set in the trainer config. No actual change to training regime here, but this is a more explicit approach
- A bug where run_B was the same run as run_A despite the correct arguments being passed has been fixed. It seems only one wandb run can be active at a time, so this is fixed by explicitly finishing reach run then re-initialising it in the attack script when it's time to log the results
- Some extra info on slurm in the scripts/README

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -152,3 +152,5 @@ results/
 
 # Slurm outputs
 slurm_logs/
```
Comment (lannelin) [inline .gitignore:154]: do we need this one now?

File: scripts/README.md [hunk 2]
```diff
@@ -56,7 +56,7 @@ Alternatively, you can call each of the scripts individually, e.g.
 sbatch trainer_scripts/drop-only_0_0_trainer.sh
 ```
 
-to train models for that particular dataset pair, where `drop-only_0_0_trainer` is one of the bash scripts generated for the `drop-only` experiment group, with dataset pair index `0` and seed index `0`. Note that logging is performed with wandb and so you will need this set up on your machine in order to proceed.
+to train models for that particular dataset pair, where `drop-only_0_0_trainer` is one of the bash scripts generated for the `drop-only` experiment group, with dataset pair index `0` and seed index `0`. Note that logging is performed with wandb and so you will need this set up on your machine in order to proceed. Note also that the output directory specified in the slurm template (see [requirements](#requirements) below.)
```
Comment (lannelin) [inline scripts/README.md:59]: revise wording of last sentence? 
"Note also <strike>that</strike> the output directory specified in the slurm template"?

File: scripts/README.md [hunk 3]
```diff
@@ -88,13 +88,13 @@ sbatch attack_scripts/drop-only_0_0_attack.sh
 
 to perform transfer attacks for that particular dataset pair, where `drop-only_0_0_trainer` is one of the bash scripts generated for the `drop-only` experiment group, with dataset pair index `0` and seed index `0`.
 
-As stated above, note that a unique run must exist on wandb for the transfer attack to be performed. Results for model A as the surrogate model (i.e. with B as the target model) will be logged to model A on wandb, and vice versa for model B
+As stated above, note that a unique run must exist on wandb for the transfer attack to be performed. Results for model A as the surrogate model (i.e. with B as the target model) will be logged to model A on wandb, and vice versa for model B.  Note also that the output directory specified in the slurm template (see [requirements](#requirements) below.)
```
Comment (lannelin) [inline scripts/README.md:91]: as above

File: scripts/attack.py [hunk 4]
```diff
@@ -26,24 +26,33 @@ def main(
...
-    )
-    model_B, run_B = download_model(
-        run_name_B,
-        entity=trainer_config["wandb"]["entity"],
-        project_name=trainer_config["wandb"]["project"],
-        version=attack_config["model_version"],
-    )
+
+    # Convinience functions to get runs + models for A and B respectively
+    def download_model_A():
```
[the remark this answers is not in the document]
Comment (lannelin) [reply scripts/attack.py:33]:
```suggestion
def download_experiment_model(model_suffix:str):
        if model_suffix not in ["A", "B"]:
            raise Exception("only A and B model suffixes are supported")
        
        return download_model(
            experiment_name=f"{experiment_pair_name}_{model_suffix}",
            entity=trainer_config["wandb"]["entity"],
            project_name=trainer_config["wandb"]["project"],
            version=attack_config["model_version"],
        )
```

requires further changes to calls to `download_model_A` and `download_model_B`

**Review by lannelin — CHANGES_REQUESTED** [review]
changes look good. I've made a couple of relatively small suggestions

File: scripts/attack.py [hunk 5]
```diff
@@ -26,24 +26,33 @@ def main(
...
+            project_name=trainer_config["wandb"]["project"],
+            version=attack_config["model_version"],
+        )
+        return model_B, run_B
+
+    # Download and prepare models - only one wandb process allowed at a time
+    model_A, run_A = download_model_A()
+    run_A.finish()
+    model_B, run_B = download_model_B()
+    run_B.finish()
```
[the remark this answers is not in the document]
Comment (philswatton) [reply scripts/attack.py:55]: Have extracted to a utility method

File: scripts/attack.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (philswatton) [reply scripts/attack.py:33]: Implementation below means this is no longer needed

File: scripts/README.md (hunk 2 above)
Comment (philswatton) [reply scripts/README.md:59]: Oops, was supposed to finish off with " must exist prior to running the scripts."! Now done

File: scripts/attack.py [hunk 6]
```diff
@@ -138,16 +147,19 @@ def main(
...
     run_A.finish()
+
+    _, run_B = download_model_B()
+    run_B.log({"B_to_A_metrics": B_to_A_metrics}, commit=True)
     run_B.finish()
 
 
 if __name__ == "__main__":
-    parser = argparse.ArgumentParser(description="Process some integers.")
+    parser = argparse.ArgumentParser(description="Process some inputs.")
```
[the remark this answers is not in the document]
Comment (philswatton) [reply scripts/attack.py]: I've fleshed these out a little while pointing to the README as the main source of truth

File: .gitignore (hunk 1 above)
Comment (philswatton) [reply .gitignore]: Not anymore - I was still referring back to some previous logs there at one point but there's now enough new ones to get rid of this. Have removed it now

**Review by joannacknight — APPROVED** [review]
I have run the tests and they all pass. I can't spot any issues in the code.

File: src/modsim2/model/load_model.py [hunk 7]
```diff
@@ -52,4 +51,44 @@ def download_model(
...
+
+
+def download_AB_models(
+    experiment_pair_name: str,
+    entity: str,
+    project_name: str,
+    version: str,
+) -> tuple[ResnetModel, ResnetModel]:
+    """
+    A function that takes as input a description of a set of two experiment
```
Comment (lannelin) [inline src/modsim2/model/load_model.py:68]: think just one pair rather than two experiment pairs?

File: src/modsim2/model/utils.py [hunk 8]
```diff
@@ -68,3 +71,37 @@ def get_run_from_name(
 
     # Return if so
     return runs[0]
+
+
+def get_run_from_name(
```
Comment (lannelin) [inline src/modsim2/model/utils.py:76]: would argue this isn't from name now? run name is constructed inside the method rather than passed as param. Could just be `get_wandb_run`?

File: src/modsim2/model/utils.py [hunk 9]
```diff
@@ -38,25 +37,29 @@ def run_exists(
...
+def _get_run_info_from_name(
     run_name: str,
     entity: str,
     project_name: str,
 ) -> wandb.Api.run:
     """
-    A convinience function for return a wandb run based on its name. Exceptions in
-    the function mean that the run name must be unique. Returns the corresponding
-    wandb run if its name is unique.
+    A convinience function for returning a wandb.API.run based on its name.
```
Comment (lannelin) [inline src/modsim2/model/utils.py:53]: v minor, sp. convenience

**Review by lannelin — APPROVED** [review]
looks good to me! A couple of minor comments to consider on naming but happy for you to take a view and then merge

File: src/modsim2/model/utils.py (hunk 8 above)
Comment (philswatton) [reply src/modsim2/model/utils.py:76]: Agreed, have changed it

File: src/modsim2/model/utils.py (hunk 9 above)
Comment (philswatton) [reply src/modsim2/model/utils.py:53]: Ah, thank you!
